proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,527 @@
+---
+
+*Disclaimer*: Proposals are point-in-time designs and decisions.
+Once approved and implemented, they become historical documents.
+If you are reading an old proposal, please be aware that the
+features described herein might have continued to evolve since.
+
+# Authenticating Users via GitHub
+
+## Problem Statement
+
+Many developers in the world have a GitHub account, and many enterprises use GitHub.
+This makes GitHub a convenient identity provider for an enterprise to use to control
+access to Kubernetes clusters for its developers.
+
+This document proposes adding a GitHubIdentityProvider resource to the Pinniped Supervisor,
+which would allow GitHub to be used as an identity provider for authentication to Kubernetes
+clusters.
+
+### How Pinniped Works Today (as of version v0.28.0)
+
+The Pinniped Supervisor currently supports OIDC-compliant identity providers and
+LDAP identity providers (including Active Directory). Unfortunately, GitHub does not offer
+either of these protocols for authenticating users. GitHub uses a slightly customized version
+of OAuth 2.0 for authenticating users, so Pinniped will need to add new features to support
+using GitHub as an identity provider.
+
+## Terminology / Concepts
+
+### GitHub OAuth 2.0 Clients
+
+For web browser-based authentication, GitHub supports two different types of OAuth 2.0 clients (compared in
+[this doc](https://docs.github.com/en/apps/creating-github-apps/about-creating-github-apps/deciding-when-to-build-a-github-app)
+and [this doc](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/differences-between-github-apps-and-oauth-apps)):
+
+- GitHub Apps (newer)
+- OAuth Apps (we will call these "GitHub OAuth Apps" below to differentiate them from the generic concept of an OAuth
+  2.0 client)
+
+Both flavors of Apps allow delegated authentication and authorization, but GitHub Apps are newer and allow
+GitHub Organization administrators to have more control over the management of the apps.
+
+Astute readers may note that OAuth 2.0 is typically used only for delegated authorization, not authentication.
+GitHub's REST API enables the use of OAuth 2.0 for authentication by offering endpoints that can be used to
+learn about the user's identity from their OAuth 2.0 access token (see below).
+
+### GitHub PATs
+
+For CLI-based (no browser required) authentication, GitHub supports two different types of
+Personal Access Tokens (PATs) (compared in
+[this doc](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens)):
+
+- Fine-grained PATs (newer)
+- Classic PATs
+
+Both flavors of PATs allow authenticating to the GitHub API, but Fine-grained PATs are newer and allow
+GitHub organization administrators to have more control over how the PAT can be used to access API resources
+related to the organization.
+
+Note that at the time of writing this document, Fine-grained PATs are a Beta feature of GitHub, and are not
```

Comment:
Do we really want to impl against a Beta feature?

Reply:
The GitHub docs recommend the use of these features, so they seem committed to keeping them.